### PR TITLE
Enable HTTP transport level tracing

### PR DESCRIPTION
**Pull request by BuddhiWathsala**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/1826

## Examples
Need to have the following configurations in the `Config.toml` to enable gRPC tracing.
```toml
[ballerina.grpc.traceLogAdvancedConfig]
# Enable printing trace logs in console
console = true              # Default is false
# Prints the trace logs to the given file
path = "testTraceLog.txt"   # Optional
# Sends the trace logs to the configured endpoint
host = "localhost"          # Optional
port = 8080                 # Optional

[ballerina.grpc.accessLogConfig]
# Enable printing access logs in console
console = true              # Default is false
# Prints the access logs to the given file
path = "testTraceLog.txt"   # Optional
```

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

File: native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+                grpcAccessLogger.setLevel(Level.INFO);
+                accessLogsEnabled = true;
+            } catch (IOException e) {
+                throw new RuntimeException("failed to setup GRPC access log file: " + filePath.getValue(), e);
+            }
+        }
+
+        if (accessLogsEnabled) {
+            System.setProperty(HTTP_ACCESS_LOG_ENABLED, "true");
+            stdErr.println("ballerina: GRPC access log enabled");
```
Comment (dilanSachi) [inline native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java:164]:
```suggestion
            stdErr.println("ballerina: gRPC access log enabled");
```

File: native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        BString logFilePath = traceLogAdvancedConfig.getStringValue(GRPC_LOG_FILE_PATH);
+        if (logFilePath != null && !logFilePath.getValue().trim().isEmpty()) {
+            try {
+                FileHandler fileHandler = new FileHandler(logFilePath.getValue(), true);
+                fileHandler.setFormatter(new GrpcTraceLogFormatter());
+                fileHandler.setLevel(Level.FINEST);
+                grpcTraceLogger.addHandler(fileHandler);
+                traceLogsEnabled = true;
+            } catch (IOException e) {
+                throw new RuntimeException("failed to setup GRPC trace log file: " + logFilePath.getValue(), e);
```
Comment (dilanSachi) [inline native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java:101]:
```suggestion
                throw new RuntimeException("Failed to setup gRPC trace log file: " + logFilePath.getValue(), e);
```

File: native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+ */
+public class GrpcLogManager extends LogManager {
+
+    static {
+        // loads logging.properties from the classpath
+        try (InputStream is = GrpcLogManager.class.getClassLoader().
+                getResourceAsStream("logging.properties")) {
+            LogManager.getLogManager().readConfiguration(is);
+        } catch (IOException e) {
+            throw new RuntimeException("failed to read logging.properties file from the classpath", e);
```
Comment (dilanSachi) [inline native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java:58]:
```suggestion
            throw new RuntimeException("Failed to read logging.properties file from the classpath", e);
```

File: native/src/test/java/io/ballerina/stdlib/grpc/logging/GrpcLogManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+import static io.ballerina.stdlib.grpc.GrpcConstants.GRPC_LOG_FILE_PATH;
+import static io.ballerina.stdlib.grpc.GrpcConstants.GRPC_TRACE_LOG_HOST;
+import static io.ballerina.stdlib.grpc.GrpcConstants.GRPC_TRACE_LOG_PORT;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+/**
+ * A unit test class for Grpc module GrpcLogManager class functions.
```
Comment (dilanSachi) [inline native/src/test/java/io/ballerina/stdlib/grpc/logging/GrpcLogManagerTest.java:49]:
```suggestion
 * A unit test class for gRPC module GrpcLogManager class functions.
```

File: native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java (hunk 1 above)
Comment (BuddhiWathsala) [reply native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java]: Fixed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/413/commits/303299de9612deb38959f15b7389ed952825454a

File: native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        if (filePath != null && !filePath.getValue().trim().isEmpty()) {
+            try {
+                FileHandler fileHandler = new FileHandler(filePath.getValue(), true);
+                fileHandler.setFormatter(new GrpcAccessLogFormatter());
+                fileHandler.setLevel(Level.INFO);
+                grpcAccessLogger.addHandler(fileHandler);
+                grpcAccessLogger.setLevel(Level.INFO);
+                accessLogsEnabled = true;
+            } catch (IOException e) {
+                throw new RuntimeException("failed to setup GRPC access log file: " + filePath.getValue(), e);
```
[the remark this answers is not in the document]
Comment (BuddhiWathsala) [reply native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java]: Fixed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/413/commits/303299de9612deb38959f15b7389ed952825454a

File: native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java (hunk 2 above)
Comment (BuddhiWathsala) [reply native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java]: Fixed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/413/commits/303299de9612deb38959f15b7389ed952825454a

File: native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java (hunk 3 above)
Comment (BuddhiWathsala) [reply native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java]: Fixed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/413/commits/303299de9612deb38959f15b7389ed952825454a

File: native/src/test/java/io/ballerina/stdlib/grpc/logging/GrpcLogManagerTest.java (hunk 4 above)
Comment (BuddhiWathsala) [reply native/src/test/java/io/ballerina/stdlib/grpc/logging/GrpcLogManagerTest.java]: Fixed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/413/commits/303299de9612deb38959f15b7389ed952825454a

File: native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to read logging.properties file from the classpath", e);
+        }
+    }
+
+    protected Logger grpcTraceLogger;
+    protected Logger grpcAccessLogger;
+
+    public GrpcLogManager(boolean traceLogConsole, BMap traceLogAdvancedConfig, BMap accessLogConfig) {
+
```
Comment (lahirusamith) [inline native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java:66]: It seems a lot of method body start with a line break as this one. Do we follow this java method writing  format?

**Comment by BuddhiWathsala** [comment]
As an offline discussion, I tried to use `HttpLogManager` directly instead of  `GrpcLogManager`, and it works as expected. One slight issue here is when a user enables gRPC tracing, it prints the following HTTP specific logs.

```sh
ballerina: HTTP trace log enabled
ballerina: HTTP access log enabled
```

IMO, it won't be an issue since the usage of HTTP as the transport layer is well known and these logs actually belong to HTTP.

@chamil321 @daneshk @dilanSachi any comment?.

File: native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java (hunk 6 above)
Comment (BuddhiWathsala) [reply native/src/main/java/io/ballerina/stdlib/grpc/logging/GrpcLogManager.java]: It is due to the automatic code styles.

**Comment by dilanSachi** [comment]
> ```shell
> trace
> ```

+1 to use the existing `HttpLogManager` for the usecase

**Comment by chamil321** [comment]
> IMO, it won't be an issue since the usage of HTTP as the transport layer is well known and these logs actually belong to HTTP.
> 
> @chamil321 @daneshk @dilanSachi any comment?.

Yeah, I also think the same. But HTTP package can add an additional param to the `initializeHttpLogs` method to pass the protocol name if the following is required.
```
ballerina: gRPC trace log enabled
ballerina: gRPC access log enabled
```

**Comment by BuddhiWathsala** [comment]
> Yeah, I also think the same. But HTTP package can add an additional param to the `HttpLogManager` constructor to pass the protocol name if the following is required.
> 
> ```
> ballerina: gRPC trace log enabled
> ballerina: gRPC access log enabled
> ```

That would be better.

**Review by daneshk — DISMISSED** [review]

**Review by daneshk — APPROVED** [review]

**Pull request merged**